From 33898cef1093f67c73f0623bae941f00034b2db9 Mon Sep 17 00:00:00 2001
From: Sven van Haastregt <sven.vanhaastregt@arm.com>
Date: Wed, 22 Jun 2022 14:03:20 +0100
Subject: [PATCH 057/172] Fix UB when validating spirv-ext option

When passing `spirv-ext=,`, SPVExt will contain empty strings.
Calling `front` on an empty string is undefined behavior, so first
ensure that the string is not empty.

(cherry picked from commit 30fb15c11f0dc5be04fddb0b3798098ca1ee0ea4)
---
 tools/llvm-spirv/llvm-spirv.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/tools/llvm-spirv/llvm-spirv.cpp b/tools/llvm-spirv/llvm-spirv.cpp
index 14d45dcc..5277712a 100644
--- a/tools/llvm-spirv/llvm-spirv.cpp
+++ b/tools/llvm-spirv/llvm-spirv.cpp
@@ -481,7 +481,8 @@ static int parseSPVExtOption(
 
   for (unsigned i = 0; i < SPVExt.size(); ++i) {
     const std::string &ExtString = SPVExt[i];
-    if ('+' != ExtString.front() && '-' != ExtString.front()) {
+    if (ExtString.empty() ||
+        ('+' != ExtString.front() && '-' != ExtString.front())) {
       errs() << "Invalid value of --spirv-ext, expected format is:\n"
              << "\t--spirv-ext=+EXT_NAME,-EXT_NAME\n";
       return -1;
-- 
2.20.1

